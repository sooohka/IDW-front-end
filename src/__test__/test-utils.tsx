/* eslint-disable @typescript-eslint/no-shadow */
/* eslint-disable no-param-reassign */
import { render, RenderOptions } from "@testing-library/react";
import React, { FC, ReactElement } from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import renderer from "react-test-renderer";
import store from "../store";
import GlobalStyle from "../style/GlobalStyle";
import theme from "../style/theme";

const AllTheProviders: FC = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <Provider store={store}>
        <BrowserRouter>
          <GlobalStyle />
          {children}
        </BrowserRouter>
      </Provider>
    </ThemeProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const thunk =
  ({ dispatch, getState }: any) =>
  (next: any) =>
  (action: any) => {
    if (typeof action === "function")
      // 여기서 반환되는 action은 thunk함수임
      return action(dispatch, getState);
    return next(action);
  };

const useStore = (stateName: string, initState: any, reducer: any) => {
  const store = {
    dispatch: jest.fn((action: any) => {
      initState = reducer(initState, action);
    }),
    getState: jest.fn(() => ({ [stateName]: initState })),
  };
  const next = jest.fn();
  const makeThunk = (action: any) => thunk(store)(next)(action);
  return { store, next, makeThunk };
};

const createWithThemeProvider = (EL: ReactElement) => {
  return renderer.create(
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {EL}
    </ThemeProvider>,
  );
};

const redux = { thunk, useStore };
export { redux, customRender as render, createWithThemeProvider };

export * from "@testing-library/react";
