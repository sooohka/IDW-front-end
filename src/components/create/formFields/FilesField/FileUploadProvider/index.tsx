import React, { createContext, useCallback, useContext, useReducer } from "react";

// actions
const DELETE_FILE = "file/deleteFile" as const;
const APPEND_FILES = "file/appendFiles" as const;
const UPDATE_FILE = "file/updateFile" as const;
const SET_ACCEPTING = "file/setIsAccepting" as const;

// action creators
const appendFiles = (files: TargetFile[]) => ({ type: APPEND_FILES, payload: { files } });
const updateFiles = (file: TargetFile) => ({ type: UPDATE_FILE, payload: { file } });
const deleteFile = (id: string) => ({ type: DELETE_FILE, payload: { id } });
const setIsAccepting = (isAccepting: boolean) => ({
  type: SET_ACCEPTING,
  payload: { isAccepting },
});

type Action =
  | ReturnType<typeof appendFiles>
  | ReturnType<typeof updateFiles>
  | ReturnType<typeof deleteFile>
  | ReturnType<typeof setIsAccepting>;

const initState: { files: TargetFile[]; isAccepting: boolean } = { files: [], isAccepting: false };
const reducer = (state = initState, action: Action): typeof initState => {
  switch (action.type) {
    case APPEND_FILES: {
      const { files } = action.payload;
      return { ...state, files: [...state.files, ...files] };
    }
    case UPDATE_FILE: {
      const { file } = action.payload;
      return { ...state, files: state.files.map((v) => (v.id === file.id ? file : v)) };
    }
    case DELETE_FILE: {
      const { id } = action.payload;
      return { ...state, files: state.files.filter((file) => file.id !== id) };
    }
    case SET_ACCEPTING: {
      const { isAccepting } = action.payload;
      return { ...state, isAccepting };
    }
    default:
      return state;
  }
};

interface FileUploadContextType {
  isAccepting: boolean;
  files: TargetFile[];
  handleFilesAppend: (files: TargetFile[]) => void;
  handleFileDelete: (id: string) => void;
  handleFileUpdate: (fils: TargetFile) => void;
  handleSetIsAccepting: (isAccepting: boolean) => void;
}

const FileUploadContext = createContext<FileUploadContextType>({
  isAccepting: false,
  files: [],
  handleFileDelete: () => {},
  handleFileUpdate: () => {},
  handleFilesAppend: () => {},
  handleSetIsAccepting: () => {},
});

const FileUploadProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initState);

  const handleFilesAppend = useCallback((files: TargetFile[]) => {
    dispatch(appendFiles(files));
  }, []);
  const handleFileDelete = useCallback((id: string) => {
    dispatch(deleteFile(id));
  }, []);
  const handleSetIsAccepting = useCallback((isAccepting: boolean) => {
    dispatch(setIsAccepting(isAccepting));
  }, []);
  const handleFileUpdate = useCallback((file: TargetFile) => {
    dispatch(updateFiles(file));
  }, []);

  const value = {
    files: state.files,
    isAccepting: state.isAccepting,
    handleFileDelete,
    handleFilesAppend,
    handleSetIsAccepting,
    handleFileUpdate,
  };
  return <FileUploadContext.Provider value={value}>{children}</FileUploadContext.Provider>;
};

const useFileUploadContext = () => useContext(FileUploadContext);

export { useFileUploadContext };
export default FileUploadProvider;
