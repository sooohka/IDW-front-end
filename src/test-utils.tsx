import { render, RenderOptions } from "@testing-library/react";
import React, { FC, ReactElement } from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import store from "./store";
import GlobalStyle from "./style/GlobalStyle";
import theme from "./style/theme";

const AllTheProviders: FC = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <Provider store={store}>
        <BrowserRouter>
          <GlobalStyle />
          {children}
        </BrowserRouter>
      </Provider>
    </ThemeProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
