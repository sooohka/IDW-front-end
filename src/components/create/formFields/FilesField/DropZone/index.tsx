import { FileRejection, useDropzone } from "react-dropzone";
import * as uuid from "uuid";
import UploadIcon from "../../../../../assets/icons/cloud-upload-alt-solid.svg";
import defaultTheme from "../../../../../style/theme";
import { useFileUploadContext } from "../FileUploadProvider";
import * as S from "./Style";

const DropZone: React.FC = () => {
  const { files, isAccepting, handleSetIsAccepting, handleFilesAppend } = useFileUploadContext();
  const onDragEnter = () => {
    handleSetIsAccepting(true);
  };

  const onDragLeave = () => {
    handleSetIsAccepting(false);
  };

  const onDrop = (acceptedFiles: File[], rejectedFiles: FileRejection[]) => {
    handleSetIsAccepting(true);
    const accFiles = acceptedFiles.map((acceptedFile) => ({
      file: acceptedFile,
      errors: [],
      id: uuid.v4(),
      isSubmitted: false,
      url: "",
    }));

    const rejFiles = rejectedFiles.map((rejectedFile) => ({
      file: rejectedFile.file,
      errors: rejectedFile.errors,
      id: uuid.v4(),
      isSubmitted: false,
      url: "",
    }));
    handleFilesAppend([...accFiles, ...rejFiles]);
    handleSetIsAccepting(false);
  };

  const handleValidation = (file: File) => {
    if (files.find((imageFile) => imageFile.file.name === file.name)) {
      return {
        code: "duplicate-file",
        message: "중복된 파일은 허용이 안됩니다.",
      };
    }
    if (file.size / 1024 / 1024 > 5)
      return {
        code: "file-too-large",
        message: `파일은 5mb를 넘을 수 없습니다`,
      };
    return null;
  };
  const { getRootProps, getInputProps } = useDropzone({
    accept: "image/*",
    maxFiles: 50,
    onDrop,
    onDragEnter,
    onDragLeave,
    validator: handleValidation,
  });
  return (
    <S.DropZone {...getRootProps()} isAccepting={isAccepting}>
      <input data-testid='dropzone-input' {...getInputProps()} />
      <UploadIcon fill={defaultTheme.colors.primary} width={50} height={50} />
      <p>Drag and Drop or click here to upload</p>
    </S.DropZone>
  );
};

export default DropZone;
