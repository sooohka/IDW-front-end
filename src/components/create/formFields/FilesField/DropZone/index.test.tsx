import { ReactElement } from "react";
import userEvent from "@testing-library/user-event";
import DropZone from "./index";
import { render, screen, act, createWithThemeProvider } from "../../../../../test/test-utils";
import { FileUploadContext } from "../FileUploadProvider";

describe("components/create/formFields/FilesField/DropZone", () => {
  let DropZoneEl: ReactElement;
  let providerValue: {
    files: TargetFile[];
    isAccepting: boolean;
    handleFileDelete: (id: string) => void;
    handleFilesAppend: (files: TargetFile[]) => void;
    handleSetIsAccepting: (isAccepting: boolean) => void;
    handleFileUpdate: (file: TargetFile) => void;
  };

  beforeEach(() => {
    DropZoneEl = <DropZone />;
    providerValue = {
      files: [],
      isAccepting: false,
      handleFileDelete: jest.fn(),
      handleFilesAppend: jest.fn(),
      handleSetIsAccepting: jest.fn(),
      handleFileUpdate: jest.fn(),
    };
  });

  it("matches snapshot", () => {
    const component = createWithThemeProvider(<DropZone />);

    expect(component.toJSON()).toMatchSnapshot();
  });
  it("upload single file", async () => {
    const file = new File(["hello"], "hello.png", { type: "image/png" });
    render(
      <FileUploadContext.Provider value={providerValue}>{DropZoneEl}</FileUploadContext.Provider>,
    );
    const inputEL = screen.getByTestId("dropzone-input") as HTMLInputElement;

    await act(async () => {
      userEvent.upload(inputEL, file);
    });

    const param = expect.arrayContaining([
      expect.objectContaining({ file: expect.objectContaining({ path: "hello.png" }) }),
    ]);
    expect(providerValue.handleFilesAppend).toBeCalledTimes(1);
    expect(providerValue.handleFilesAppend).toHaveBeenCalledWith(param);
    expect(providerValue.handleSetIsAccepting).toBeCalledTimes(2);
  });

  it("upload multiple files", async () => {
    const files = [
      new File(["blobbbbba"], "apple.jpeg", { type: "image/jpeg" }),
      new File(["blobbbbbb"], "banana.png", { type: "image/png" }),
    ];
    render(
      <FileUploadContext.Provider value={providerValue}>{DropZoneEl}</FileUploadContext.Provider>,
    );
    const inputEL = screen.getByTestId("dropzone-input") as HTMLInputElement;

    await act(async () => {
      userEvent.upload(inputEL, files);
    });

    const param = expect.arrayContaining([
      expect.objectContaining({ file: expect.objectContaining({ path: "apple.jpeg" }) }),
      expect.objectContaining({ file: expect.objectContaining({ path: "banana.png" }) }),
    ]);
    expect(providerValue.handleFilesAppend).toBeCalledWith(param);
    expect(providerValue.handleFilesAppend).toBeCalledTimes(1);
    expect(providerValue.handleSetIsAccepting).toBeCalledTimes(2);
  });

  it("upload file size above 5mb", async () => {
    const file = new File(["blob"], "hi.png", { type: "image/png" });
    Object.defineProperty(file, "size", { value: 1024 * 1024 * 6 });
    render(
      <FileUploadContext.Provider value={providerValue}>{DropZoneEl}</FileUploadContext.Provider>,
    );
    const inputEL = screen.getByTestId("dropzone-input") as HTMLInputElement;

    await act(async () => {
      userEvent.upload(inputEL, file);
    });

    const param = expect.arrayContaining([
      expect.objectContaining({
        errors: expect.arrayContaining([expect.objectContaining({ code: "file-too-large" })]),
        file: expect.objectContaining({
          path: "hi.png",
        }),
      }),
    ]);
    expect(providerValue.handleFilesAppend).toBeCalledWith(param);
    expect(providerValue.handleFilesAppend).toBeCalledTimes(1);
    expect(providerValue.handleSetIsAccepting).toBeCalledTimes(2);
  });

  it("upload not supported file with file size above 5mb", async () => {
    const file = new File(["pdf blob"], "qwe.pdf", { type: "application/pdf" });
    Object.defineProperty(file, "size", { value: 1024 * 1024 * 6 });
    render(
      <FileUploadContext.Provider value={providerValue}>{DropZoneEl}</FileUploadContext.Provider>,
    );
    const inputEl = screen.getByTestId("dropzone-input") as HTMLInputElement;

    await act(async () => {
      userEvent.upload(inputEl, file);
    });

    const param = expect.arrayContaining([
      expect.objectContaining({
        file: expect.objectContaining({ path: "qwe.pdf" }),
        errors: expect.arrayContaining([
          expect.objectContaining({ code: "file-invalid-type" }),
          expect.objectContaining({ code: "file-too-large" }),
        ]),
      }),
    ]);
    expect(providerValue.handleFilesAppend).toBeCalledWith(param);
    expect(providerValue.handleSetIsAccepting).toBeCalledTimes(2);
  });
});
