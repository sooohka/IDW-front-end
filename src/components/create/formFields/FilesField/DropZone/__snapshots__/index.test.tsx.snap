// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/create/formFields/FilesField/DropZone matches snapshot 1`] = `
<div
  className="sc-gsDKAQ bFxvOH"
  onBlur={[Function]}
  onClick={[Function]}
  onDragEnter={[Function]}
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDrop={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <input
    accept="image/*"
    autoComplete="off"
    data-testid="dropzone-input"
    multiple={true}
    onChange={[Function]}
    onClick={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    tabIndex={-1}
    type="file"
  />
  <svg
    fill="#2d3436"
    height={50}
    width={50}
  />
  <p>
    Drag and Drop or click here to upload
  </p>
</div>
`;
